### DISCUSSION TITLE
How to get pybricks-micropython executable v3.2 (or v3.x) for EV3?

### DISCUSSION THREAD

**Discussion by AAATom** [pybricks/support#1075] [Q&A] [answered]
I've spent several hours today trying to get my hands on the mysterious EV3 "pybricks-micropython executable" of any 3.x version which I could then, supposedly, _"install in `/usr/local/bin/` on EV3 brick"_ to try a newer version than the ubiquitous v2.0 we already use - all in vain.

I've found several old posts that even contain links to the "pybricks-micropython executable" zip file but they have all expired long ago.
None of the releases [here](https://github.com/pybricks/pybricks-micropython/releases) appears to have it either.

**I would appreciate it very much if someone knowledgeable could point me to a place from where I could download one, preferably for the latest stable release** (should be 3.2, right?).

Background: I'm co-coaching my son's school team which is preparing for national WRO competition. The v2.0 works fine up to its limits, one of largest being that `DriveBase` has `Stop.HOLD` as default and no way to change it (like using `then` parameter in v3.x), what makes movements jerky. I assume this could be worked around with some clever code but, hey, they are just kids starting with Python and "robotics" and would be happy to see such a thing working out-of -the-box (like they found it in the docs, alas, for v3.2).

**Comment by dlech** [comment, marked as answer]
Up to now, the 3.x versions have been focused on Powered Up hubs and EV3 hasn't been extensively tested so v3.2 is not guaranteed to be working 100% properly on EV3. We've been considering getting EV3 back to supported status with the next release (v3.3), so it would be helpful if you could try the latest builds and let us know how it goes. You can always find the most recent build here: https://nightly.link/pybricks/pybricks-micropython/workflows/build/master

**Reply by jaguilar** [reply]
For anyone finding this answer after August '23, make sure to check out #1035 before proceeding.

**Comment by AAATom** [comment]
Wow, that was fast! :)
Thank you!

I know it might not work 100% as it is now but it could have more benefits than drawbacks - we'll see. Good to know EV3 support could be coming back!
I'll try it tomorrow and let you know.

**Comment by AAATom** [comment]
David,

how do I get permissions to write to `/usr/local/bin/` on the EV3 brick?

**Reply by dlech** [reply]
Use `sudo`.

**Reply by AAATom** [reply]
Yes, thanks, I managed to do it!

I had some issues but found a workaround. Will post it next.

**Comment by AAATom** [comment]
So, I had problems with `scp` when running it on EV3 (via SSH terminal) - it would just hang without any output. I did connect EV3 to my network and I could ping it from EV3 but `scp` just hang.
To get around it I run `pscp` (PuTTY's scp command) on my Windows machine to copy "pybricks-micropython" file to `/home/robot/` on EV3. Then I `sudo cp`-ed it to `/usr/local/bin/`.
However, that did not make it run yet - I had to change permissions for "pybricks-micropython" file to allow execution.

### All in one place for Linux beginners like me:
**0.** Make sure you can connect to your EV3. 
I used USB cable to connect EV3 to my machine running Windows.
If you can download and run programs from VS Code to your EV3 then you are all set up.
I used SSH Terminal from VS Code - right-click on the "ev3dev" in ev3dev Device Browser and choose "Open SSH Terminal". It does all authentication automatically and you will get the prompt at `/home/robot/`.

**1.** Install [PuTTY](https://www.putty.org/) - it will add `pscp` command you need to transfer "pybricks-micropython" file.

**2.** Open command prompt in folder where you have the "pybricks-micropython" file.

**3.** Run this command at the prompt to copy "pybricks-micropython" file to EV3:
        `pscp -scp pybricks-micropython robot@ev3dev:/home/robot/`
It should take couple of seconds and it will report successful copying.

**4.** Open SSH terminal and connect to EV3 (easiest is to do it from VS Code - see above).

**5.** Run this command in terminal to copy "pybricks-micropython" file to `/usr/local/bin/` directory:
       `sudo cp /home/robot/pybricks-micropython /usr/local/bin/`

**6.** Run this command in terminal to change directory to `/usr/local/bin/`:
       `cd /usr/local/bin/`

**7.** Run this command in terminal to make "pybricks-micropython" file executable:
       `sudo chmod +x pybricks-micropython`

**8.** Reboot EV3
      This may not be necessary  - I rebooted before step 7 so it did not work but it worked right away when I changed "pybricks-micropython" file to executable.

To get the version of pybricks you can use `pybricks.version` property:
```
from pybricks import version
print(version)
```
You will get something like this:
> ('ev3', '3.3.0b5', 'v1.20.0-23-g6c633a8dd on 2023-05-19')

**Comment by BossieL** [comment]
Is there any news regarding support for EV3 with v3.2 (or even 3.3)?
- when will 3.3 be released? will it automatically include support for EV3? 
- how do we flash the SD card? 
- will the libraries in Visual Studio Code also be updated automatically? or is there a new extension to be installed? 

@AAATom 
You described a solution on how to upgrade the EV3 brick with v3.2. 
I checked the link with drivers, but which one do I need to use? 
Does the v3.2 work as expected? do all methods work of DriveBase? 
In which editor do you perform the coding as Visual Source Code only has the v2.0 extension.

**Reply by dlech** [reply]
> when will 3.3 be released?

Some time this year.

> will it automatically include support for EV3?

There are some known issues, but hopefully we can find time to sort them out for this release.

> how do we flash the SD card?

The same way as always. https://pybricks.com/ev3-micropython/startinstall.html

Then connect the EV3 to to the internet and `sudo apt update && sudo apt install pybricks-micropython3`.

You can already do this today. 

> will the libraries in Visual Studio Code also be updated automatically? or is there a new extension to be installed?

The same extension works. Also see https://pybricks.com/projects/tutorials/dev/tools/vscode/

**Reply by BossieL** [reply]
Hi dlech, thanks for your swift response. 
As I already have v2.0 on my SD Card and which is working, I connected in Visual Studio to the Brick, opened a shell and launched your command with sudo. 
This is the answer I receive: 

```console
robot@ev3dev:~$ cd /home/robot
robot@ev3dev:~$ sudo apt update && sudo apt install pybricks-micropyth
Err:1 http://httpredir.debian.org/debian stretch InRelease                                                    
  Could not resolve 'httpredir.debian.org'
Err:2 http://security.debian.org stretch/updates InRelease                                                    
  Could not resolve 'security.debian.org'
Err:3 http://archive.ev3dev.org/debian stretch InRelease                 
  Could not resolve 'archive.ev3dev.org'
Reading package lists... Done        
Building dependency tree       
Reading state information... Done
5 packages can be upgraded. Run 'apt list --upgradable' to see them.
W: Failed to fetch http://httpredir.debian.org/debian/dists/stretch/InRelease  Could not resolve 'httpredir.debian.org'
W: Failed to fetch http://security.debian.org/dists/stretch/updates/InRelease  Could not resolve 'security.debian.org'
W: Failed to fetch http://archive.ev3dev.org/debian/dists/stretch/InRelease  Could not resolve 'archive.ev3dev.org'
W: Some index files failed to download. They have been ignored, or old ones used instead.
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package pybricks-micropyth
```
Any further suggestions? It seems not to find the updates on the internet and then it throws naturally some other errors.

**Reply by dlech** [reply]
The errors with `http://httpredir.debian.org/debian` not working any more is a known issue and can be ignored. But since it also can't find `http://archive.ev3dev.org/debian`, it sounds like the EV3 is not connected to the internet.

Also, there is a typo in the package name, it should be `pybricks-micropython3`, not `pybricks-micropyth`.

**Reply by BossieL** [reply]
Ok. So the commands I initiate effectively asks the Brick to connect itself to the internet. I have no wifi dongle on this brick, so obviously it fails (and makes sense). 
The method use by AAATom seems to me an option, but then I have to find the correct libraries and then perform a copy. 
The link  
When I click on the link "https://nightly.link/pybricks/pybricks-micropython/workflows/build/master" it shows a lot of versions, but I have no clue what to download and copy.  
Sorry for all those questions. :( 
Very much appreciated for this help. Kr.

**Reply by dlech** [reply]
There are other ways to connect to the internet: https://www.ev3dev.org/docs/networking/

In the nightly builds, the one that starts with `pybricks-micropython` is for EV3.
